contracts/riskon/UniswapV3RiskOnVault.sol

```diff
@@ -0,0 +1,673 @@
+    function balanceOfPoolWants(MintInfo memory _mintInfo) internal view returns (uint256, uint256) {
+        if (_mintInfo.tokenId == 0) return (0, 0);
+        (uint160 _sqrtPriceX96, , , , , ,) = pool.slot0();
+        return LiquidityAmounts.getAmountsForLiquidity(_sqrtPriceX96, TickMath.getSqrtRatioAtTick(_mintInfo.tickLower), TickMath.getSqrtRatioAtTick(_mintInfo.tickUpper), balanceOfLpToken(_mintInfo.tokenId));
+    }
+
+    /// @notice Harvests the Strategy
+    /// @return  _rewardsTokens The reward tokens list
+    /// @return _claimAmounts The claim amounts list
+    function harvest() public override returns (address[] memory _rewardsTokens, uint256[] memory _claimAmounts) {
+        lastHarvest = block.timestamp;
+        _rewardsTokens = new address[](2);
+        _rewardsTokens[0] = token0();
+        _rewardsTokens[1] = token1();
+        _claimAmounts = new uint256[](2);
+        uint256 _amount0;
+        uint256 _amount1;
+        if (baseMintInfo.tokenId > 0) {
+            (_amount0, _amount1) = __collectAll(baseMintInfo.tokenId);
+            _claimAmounts[0] += _amount0;
+            _claimAmounts[1] += _amount1;
+        }
+        if (limitMintInfo.tokenId > 0) {
+            (_amount0, _amount1) = __collectAll(limitMintInfo.tokenId);
+            _claimAmounts[0] += _amount0;
+            _claimAmounts[1] += _amount1;
+        }
+
+        if (profitFeeBps > 0 && address(treasury) != address(0)) {
+            if (_claimAmounts[0] > 0) {
+                uint256 claimAmount0Fee = _claimAmounts[0] * profitFeeBps / 10000;
+                _claimAmounts[0] -= claimAmount0Fee;
+                IERC20Upgradeable(token0()).safeApprove(address(treasury), 0);
+                IERC20Upgradeable(token0()).safeApprove(address(treasury), claimAmount0Fee);
+                treasury.receiveProfitFromVault(token0(), claimAmount0Fee);
+            }
+            if (_claimAmounts[1] > 0) {
+                uint256 claimAmount1Fee = _claimAmounts[1] * profitFeeBps / 10000;
+                _claimAmounts[1] -= claimAmount1Fee;
+                IERC20Upgradeable(token1()).safeApprove(address(treasury), 0);
+                IERC20Upgradeable(token1()).safeApprove(address(treasury), claimAmount1Fee);
+                treasury.receiveProfitFromVault(token1(), claimAmount1Fee);
+            }
+        }
+        emit StrategyReported(_rewardsTokens, _claimAmounts);
+    }
+
+    /// @notice Lend
+    /// @param _amount The amount of lend
+    function lend(uint256 _amount) external isOwner whenNotEmergency nonReentrant override {
+        IERC20Upgradeable(wantToken).safeTransferFrom(msg.sender, address(this), _amount);
+
+        if (manageFeeBps > 0 && address(treasury) != address(0)) {
+            uint256 manageFee = _amount * manageFeeBps / 10000;
+            IERC20Upgradeable(wantToken).safeApprove(address(treasury), 0);
+            IERC20Upgradeable(wantToken).safeApprove(address(treasury), manageFee);
```

Comment:
can put L231 and 232 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(token0()).safeApprove(address(treasury), type(uint256).max);
```